### PR TITLE
connect: update supported envoy versions to 1.18.2, 1.17.2, 1.16.3, and 1.15.4

### PR DESCRIPTION
**Pull request by rboyer**
The only thing that needed fixing up pertained to this section of the [1.18.x release notes](https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.18.0):

> grpc_stats: the default value for stats_for_all_methods is switched from true to false, in order to avoid possible memory exhaustion due to an untrusted downstream sending a large number of unique method names. The previous default value was deprecated in version 1.14.0. This only changes the behavior when the value is not set. The previous behavior can be used by setting the value to true. This behavior change by be overridden by setting runtime feature envoy.deprecated_features.grpc_stats_filter_enable_stats_for_all_methods_by_default.

For now to maintain status-quo I'm explicitly setting `stats_for_all_methods=true` in all versions to avoid relying upon the default.

Additionally the naming of the emitted metrics for these gRPC requests changed slightly so the integration test assertions for `case-grpc` needed adjusting.

### CODE REVIEW

File: agent/xds/clusters_test.go [hunk 1]
```diff
@@ -642,6 +643,7 @@ func TestClustersFromSnapshot(t *testing.T) {
 	}
 
 	latestEnvoyVersion := proxysupport.EnvoyVersions[0]
+	latestEnvoyVersion_v2 := proxysupport.EnvoyVersionsV2[0]
```
Comment (rboyer) [inline agent/xds/clusters_test.go:646]: I realized when I went to update these golden tests that my last refactor hadn't accounted for skipping `v2-compat` tests for `>= 1.17.x`. All of the golden tests got similar adjustments.

File: agent/xds/testdata/listeners/connect-proxy-with-chain-and-overrides.envoy-1-18-x.golden [hunk 2]
```diff
@@ -28,6 +28,13 @@
                   "routeConfigName": "db"
                 },
                 "httpFilters": [
+                  {
```
Comment (rboyer) [inline agent/xds/testdata/listeners/connect-proxy-with-chain-and-overrides.envoy-1-18-x.golden:31]: an actual difference

File: agent/xds/testdata/listeners/connect-proxy-with-chain-and-overrides.v2compat.envoy-1-16-x.golden [hunk 3]
```diff
@@ -28,6 +28,13 @@
                   "routeConfigName": "db"
                 },
                 "httpFilters": [
+                  {
```
Comment (rboyer) [inline agent/xds/testdata/listeners/connect-proxy-with-chain-and-overrides.v2compat.envoy-1-16-x.golden:31]: an actual difference

File: website/content/docs/connect/proxies/envoy.mdx [hunk 4]
```diff
@@ -34,10 +34,10 @@ compatible Envoy versions.
...
 | Consul Version      | Compatible Envoy Versions        |
 | ------------------- | -------------------------------- |
-| 1.10.x              | 1.17.0, 1.16.2, 1.15.3, 1.14.6‡, |
-| 1.9.x               | 1.16.2, 1.15.3, 1.14.6‡, 1.13.7‡ |
-| 1.8.x               | 1.14.6, 1.13.7, 1.12.7, 1.11.2   |
-| 1.7.x               | 1.13.7, 1.12.7, 1.11.2, 1.10.0\* |
+| 1.10.x              | 1.18.2, 1.17.2, 1.16.3, 1.15.4   |
+| 1.9.x               | 1.16.3, 1.15.4, 1.14.7‡, 1.13.8‡ |
+| 1.8.x               | 1.14.7, 1.13.8, 1.12.7, 1.11.2   |
+| 1.7.x               | 1.13.8, 1.12.7, 1.11.2, 1.10.0\* |
```
Comment (rboyer) [inline website/content/docs/connect/proxies/envoy.mdx:40]: The manual backports of this pr to the older series will finish making these 3 lines true.

File: agent/xds/listeners.go [hunk 5]
```diff
@@ -1588,6 +1589,24 @@ func makeHTTPFilter(opts listenerFilterOpts) (*envoy_listener_v3.Filter, error)
 		cfg.HttpFilters = append([]*envoy_http_v3.HttpFilter{{
 			Name: "envoy.filters.http.grpc_http1_bridge",
 		}}, cfg.HttpFilters...)
+
+		// In envoy 1.14.x the default value "stats_for_all_methods=true"e was
```
Comment (freddygv) [inline agent/xds/listeners.go:1593]:
```suggestion
		// In envoy 1.14.x the default value "stats_for_all_methods=true" was
```

**Review by freddygv — COMMENTED** [review]
Do we use `test-envoy-versions.sh` anywhere? Just saw it wasn't updated, but maybe can be deleted.

**Comment by rboyer** [comment]
@freddygv deleting the debugging file `test-envoy-versions.sh`

**Review by freddygv — APPROVED** [review]

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/360035.

**Comment by hc-github-team-consul-core** [comment]
:cherries::white_check_mark: Cherry pick of commit abc1dc0fe958e2ab82cb90aab97cbe46be0bc101 onto `release/1.10.x` succeeded!

**Pull request merged**